### ISSUE TITLE
Rule Change: `no-promise-executor-return` should allow arrow functions with no `{...}`

### ISSUE THREAD

**Issue by nopeless** [eslint/eslint#17278]
### What rule do you want to change?

`no-promise-executor-return

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

Change default behavior

### Example code

```js
// case 1
const sleep = (t) => new Promise(r => setTimeout(r, t))
const apiPromise = new Promise(r => apiRequestWithCallback("...", r));

// case 2
let getResolver;
// might be frowned upon
new Promise(r => j = r);
```

### What does the rule currently do for this code?

does not allow this syntax because it is returning a value

### What will the rule do after it's changed?

for case 1. require `void` in front of the arrow function

for case 2. ignore assignment returns

I believe these should be the default (not an option proposal)

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @nopeless, thanks for the issue! 

This was already discussed in https://github.com/eslint/eslint/issues/13668. I'd personally only support adding an option that allows `=> void <expression>` and `return void <expression>`. Curios what other team members think about this.

**Comment by nopeless** [comment]
> Hi @nopeless, thanks for the issue!
> 
> This was already discussed in #13668. I'd personally only support adding an option that allows `=> void <expression>` and `return void <expression>`. Curios what other team members think about this.

Oops, I did search for this rule proposal, I accidentally missed it, sorry

I don't see how case 2 isn't in your favor though, I'd like to hear your thoughts

Thanks for looking into the issue

**Comment by nopeless** [comment]
> > I don't see how case 2 isn't in your favor though, I'd like to hear your thoughts
> 
> I don't think this rule should be deciding which patterns look like intent to return a value and which don't. `void` as an indication that there is no intent to return a value seems fine to me.

Maybe my thinking is wrong, but I don't think there exists a developer smart enough to know that the assignment operator in js also returns the LHS value but ignorant enough to think that the function's return value has some sort of meaning.

I did see some new developers struggling to debug their code because they used `if (a = b)`. I personally use `let result; if (result = someFunction()) {}` pattern a lot. (this is prevented by `no-cond-assign`)

However, I don't mind putting `{}` around assignments. Its a rather rare case to do so

**Comment by nopeless** [comment]
> > This was already discussed in #13668. I'd personally only support adding an option that allows => void  and return void . Curios what other team members think about this.
> 
> I'm 👍 for this to cover case 1. It's a clear signal that you know you're not returning a value.
> 
> I'm 👎 for covering case 2.
> 
> @nopeless do you want to submit a pull request for that change?

I'd sure love to try, I'll look at the contributing guide

**Comment by nopeless** [comment]
@nzakas @mdjermanovic

**Comment by nopeless** [comment]
Meanwhile, this seems like a similar issue to `array-callback-return` rule. Should I submit a different issue for that?

**Comment by nopeless** [comment]
@mdjermanovic I'd like your opinion on this

**Comment by mdjermanovic** [comment]
> @mdjermanovic I'd like your opinion on this

On allowing `.forEach(() => void...)` in `array-callback-return`? Yes, that seems similar and we can discuss it, and yes it would be better to submit a separate issue.
